include/pybind11/detail/type_caster_base.h

```diff
@@ -852,10 +866,15 @@ struct is_copy_constructible<std::pair<T1, T2>>
 template <typename T, typename SFINAE = void>
 struct is_copy_assignable : std::is_copy_assignable<T> {};
 template <typename Container>
-struct is_copy_assignable<Container,
-                          enable_if_t<all_of<std::is_copy_assignable<Container>,
-                                             std::is_same<typename Container::value_type &,
-                                                          typename Container::reference>>::value>>
+struct is_copy_assignable<
+    Container,
+    enable_if_t<
+        all_of<std::is_copy_assignable<Container>,
+               std::is_same<typename Container::value_type &, typename Container::reference>,
+               // Avoid infinite recursion (list types)
+               negation<std::is_same<Container, typename Container::value_type>>,
```

Comment:
We also have `none_of` in addition to all_of and negation if that would make the logic cleaner here.